### PR TITLE
Fix geo tile bounding boxes to be consistent with arithmetic method

### LINKED ISSUE #92611: [CI] GeoGridQueryBuilderTests testBoundingBoxQuantize failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/p4t5iiy4zebjw/tests/:x-pack:plugin:spatial:test/org.elasticsearch.xpack.spatial.index.query.GeoGridQueryBuilderTests/testBoundingBoxQuantize

**Reproduction line:**
```
gradlew ':x-pack:plugin:spatial:test' --tests "org.elasticsearch.xpack.spatial.index.query.GeoGridQueryBuilderTests.testBoundingBoxQuantize" -Dtests.seed=645E29000141A52C -Dtests.locale=ar-KW -Dtests.timezone=America/Argentina/Jujuy -Druntime.java=19
```

**Applicable branches:**
main

**Reproduces locally?:**
Yes

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.spatial.index.query.GeoGridQueryBuilderTests&tests.test=testBoundingBoxQuantize

**Failure excerpt:**
```
java.lang.AssertionError: expected:<false> but was:<true>

  at __randomizedtesting.SeedInfo.seed([645E29000141A52C:E15FA09F68360DE5]:0)
  at org.junit.Assert.fail(Assert.java:88)
  at org.junit.Assert.failNotEquals(Assert.java:834)
  at org.junit.Assert.assertEquals(Assert.java:118)
  at org.junit.Assert.assertEquals(Assert.java:144)
  at org.elasticsearch.xpack.spatial.index.query.GeoGridQueryBuilderTests.assertBoundingBox(GeoGridQueryBuilderTests.java:298)
  at org.elasticsearch.xpack.spatial.index.query.GeoGridQueryBuilderTests.testBoundingBoxQuantize(GeoGridQueryBuilderTests.java:290)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  at java.lang.reflect.Method.invoke(Method.java:578)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1589)

```

### LINKED ISSUE #95574: [CI] GeoGridAggAndQueryConsistencyIT testGeoShapeGeoTile failing
Started failing today: https://gradle-enterprise.elastic.co/scans/tests?search.startTimeMax=1682510196670&search.startTimeMin=1681851600000&search.timeZoneId=Europe/Bucharest&tests.container=org.elasticsearch.xpack.spatial.search.GeoGridAggAndQueryConsistencyIT&tests.test=testGeoShapeGeoTile

**Build scan:**
https://gradle-enterprise.elastic.co/s/qddijdieojw6q/tests/:x-pack:plugin:spatial:internalClusterTest/org.elasticsearch.xpack.spatial.search.GeoGridAggAndQueryConsistencyIT/testGeoShapeGeoTile

**Reproduction line:**
```
./gradlew ':x-pack:plugin:spatial:internalClusterTest' --tests "org.elasticsearch.xpack.spatial.search.GeoGridAggAndQueryConsistencyIT.testGeoShapeGeoTile" -Dtests.seed=24AC435F8381E2D0 -Dtests.locale=fr -Dtests.timezone=Europe/Ljubljana -Druntime.java=20
```

**Applicable branches:**
main

**Reproduces locally?:**
Yes

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.spatial.search.GeoGridAggAndQueryConsistencyIT&tests.test=testGeoShapeGeoTile

**Failure excerpt:**
```
java.lang.AssertionError: Expected hits at precision 26 for H3 cell 26/0/33554410
Expected: <15L>
     but: was <14L>

  at __randomizedtesting.SeedInfo.seed([24AC435F8381E2D0:F24F5D998EB839F0]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.elasticsearch.xpack.spatial.search.GeoGridAggAndQueryConsistencyIT.assertQuery(GeoGridAggAndQueryConsistencyIT.java:316)
  at org.elasticsearch.xpack.spatial.search.GeoGridAggAndQueryConsistencyIT.doTestGrid(GeoGridAggAndQueryConsistencyIT.java:279)
  at org.elasticsearch.xpack.spatial.search.GeoGridAggAndQueryConsistencyIT.doTestGeotileGrid(GeoGridAggAndQueryConsistencyIT.java:202)
  at org.elasticsearch.xpack.spatial.search.GeoGridAggAndQueryConsistencyIT.testGeoShapeGeoTile(GeoGridAggAndQueryConsistencyIT.java:84)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  at java.lang.reflect.Method.invoke(Method.java:578)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1623)

```

### PR DESCRIPTION
**Pull request by iverase**
In Elasticsearch we sometimes use geotile grid using arithmetic operations, for example calling GeoTileUtils#getXTile nd sometimes we resolve things geometrically by calling  GeoTileUtils#toBoundingBox. It is easy to prove that the geometric solution is equivalent to the arithmetic solution until zoom 19 using lucene encoding scheme for geo. After that, we can observe instabilities and solutions are not consistent.

In order to fix that we added some logic to GeoTileUtils#toBoundingBox that make sure the generated bounding boxes are consistent with the arithmetic solution and hence we can use both approaches indistintable, regardless the zoom level.

fixes https://github.com/elastic/elasticsearch/issues/92611
fixes https://github.com/elastic/elasticsearch/issues/95574

### CODE REVIEW

**Comment by iverase** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by iverase** [comment]
hit https://github.com/elastic/elasticsearch/issues/92058, it is being muted.

@elasticmachine run elasticsearch-ci/bwc

**Comment by iverase** [comment]
@elasticmachine update branch

**Review by craigtaverner — APPROVED** [review]
LGTM, but did have a couple of small comments.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileUtils.java [hunk 1]
```diff
@@ -256,10 +265,29 @@ private static double tileToLon(double xTile, double tiles) {
...
+        // the idea here is that the latitude returned belongs to the tile and the next latitude up belongs to the next tile
+        // therefore we can be in the current tile and we need to find the point up just before the next tile,
+        // or we are in the other tile and we need to find the first point down that belong to this tile.
+        return findBoundaryPoint(qLat, computedYTile, tiles, computedYTile == yTile ? LUCENE_LAT_RES : -LUCENE_LAT_RES);
+    }
+
+    private static double findBoundaryPoint(double qLat, int yTile, int tiles, double step) {
+        final double nextQLat = qLat + step;
+        final int nextYTile = getYTile(nextQLat, tiles);
+        if (yTile != nextYTile) {
```
[the remark this answers is not in the document]
Comment (iverase) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileUtils.java:284]: You need to be aware that numbering for latitude goes from positive to negative.

 If we go up (steps are positive) we are already in the right tile and we want to find the point where the tile changes. When the next point is in a different tile, the current point is the one we are looking for.

If we go down (steps are negative) then we are in the wrong tile so we need to find the point where it changes. That next point where it changes is the one we are looking for.

makes more sense?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileBoundedPredicate.java [hunk 2]
```diff
@@ -35,15 +38,15 @@ public GeoTileBoundedPredicate(int precision, GeoBoundingBox bbox) {
             final int minY = GeoTileUtils.getYTile(bbox.top(), tiles);
             final Rectangle minTile = GeoTileUtils.toBoundingBox(minX, minY, precision);
             // touching tiles are excluded, they need to share at least one interior point
-            this.leftX = minTile.getMaxX() == bbox.left() ? minX + 1 : minX;
-            this.minY = minTile.getMinY() == bbox.top() ? minY + 1 : minY;
+            this.leftX = quantizeLon(minTile.getMaxX()) == quantizeLon(bbox.left()) ? minX + 1 : minX;
```
[the remark this answers is not in the document]
Comment (iverase) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileBoundedPredicate.java:41]: yes, updated.

**Pull request merged**